[Roach] 미션3: 체스판 초기화

tmdgusya opened the pull request:
이번에는 미션3 에 있는 **구분자 하나마다 Commit** 을 등록했다. 
**구현은 feat  / 리팩토링은 refactor** 이런식으로, 이렇게 하니까 훨씬 커밋 내역이 명확해지는것 같았다.
**그래도 아직 부족한 부분이 있을지 모르니 리뷰어 분께서 부족하신 부분은 알려주시면 감사하겠습니다 :)**

요구사항에 최대한 준수하였고, 타입에 관한것은 엄격하게 해야한다고 생각하여 Representation 의 경우는 따로 Enum 으로 분리하였다.

tmdgusya committed:
feat Pawn 색에 따른 출력 문자 부여

Pawn.java

  - Representation 을 char 가 아닌 따로 Enum Class 를 이용한 이유는 타입의 유효성을 위해 설정하였음. 예를 들면 'c' 가 들어온다해도 현재 요구사항에서는 막을 방법이 if 를 이용한 validate 말고는 없기에, Enum 을 이용하여 타입에 대한 제한을 걸어둠.

  - 생성자중 representation 을 받는 생성자에 this(color) 로 한 이유는 Client 의 실수로, representation 이 Black 인데도 색깔은 white 일수도 있기에, 이를 방지하기 위해 색깔에 맞는 값을 넣어주는 이전 생성자를 이용함.

  - getRepresentation() 에서 name() 을 리턴하는 이유는 후에 사용될 Board 에서의 이름 출력을 고려하여 아래와 같이 설계함.

 PawnName.java

  - 각 색깔 속성에 맞는 representation 값을 가지도록 하였음.
  - getRepresentation 을 통해서 해당 representation 값을 얻을 수 있음.

tmdgusya committed:
test PawnTest 새로운 메소드 테스트 추가

- 색깔에 따른 Representation 에 대한 검증도 필요하므로, 각 테스트에 getRepresentation() 을 넣어 유효성을 검증할 수 있도록 하였음.

- 유효하지 않은 Representation 을 넣어도 보정이 되는지도 테스트를 추가하였음.

tmdgusya committed:
feat 체스판 초기화

Board.java
 - BOARD_SIZE = 8 , 반복문에서 사용할때 가독성을 위해 상수로 추가
 - 요구사항에 따른 색깔별 List 를 추가하고, initialize() 에서 그 구현체를 넣어주도록 했음.
 - initialize() 체스판을 초기화 하는 역할으로, 하얀색폰과 검은색폰을 8개씩 추가해줌.
 - add() 의 경우는 Color 로 구분할 수 있기에 Color 에 따른 List 를 골라서 add 하도록 함.
 - size() 의 경우 각 Color 에 따른 size() 를 리턴할 수 있도록 메소드 두개 생성. (findPawn 과 getPawnResult 도 마찬가지)
 - getPawnsResult() -> 문자열 연산이 많기에 StringBuilder 를 이용하였음

tmdgusya committed:
test BoardTest 보드 상태 테스트 추가

- checkPawnCount : 보드가 초기화 되고나면, Pawn 들이 정상적으로 8개씩 등록되어 지는지 확인하는 테스트를 추가하였음
- checkPawnRepresentationOnTheBoard : 보드에 올려진 폰들의 Representation이 합쳐진 것들이 정상적으로 출력되는지 확인하는 테스트를 추가하였음.

tmdgusya committed:
refactor Board Class 중복 리팩토링 제거

Board.java

- 기존 add 클래스는 보드에 pawn 을 추가하는 역할이지만, 유효한 리스트 마저 찾은뒤 add 하는 역할까지 책임지고 있었음, 이를 getValidateInputList 메소드를 통해 책임을 분리하였음.

- 기존 intialize() 는 List()에 구현체를 넣는 역할과 Pawn 을 보드 사이즈만큼 추가하는 역할을 가지고 있었으나, 해당 Pawn 을 추가하는 기능을 addPawnToBoardSize() 로 분리하여 책임을 분리하였음.

BoardTest.java

- intialize() 로 구현체를 주입하다보니, 테스트 코드에 대한 수정도 필요하여 테스트 코드도 수정하였음

tmdgusya committed:
feat Print 메소드 구현

Board.java

- 현재 보드판위의 상태를 나타내는 Print 메소드를 구현하기 위해 Board 판위의 상태를 String 문자열로 바꿔내는 Method 인 getBoardStatusToString 를 추가하였음, 보드 판위의 BLANK 즉 공백 란을 나타내기 위해 BLANK 상수를 추가하였고, 초기 공백 구간이라는 뜻의 initialSpaceInterval 상수값도 추가하였음.
- print 를 통해서 Console 에 Print() 할 수 있음.
- private 메소드 는 제일 아래있는게 깔끔해보여서 아래로 이동시킴

BoardTest.java

- getBoardStatusToString 가 정확히 Board 의 상태를 나타내는지 유효성을 검증하기 위해 checkBoardPrintMessage() 생성

tmdgusya committed:
feat main() 구현

- userInputCommand 를 초기에 한번 받고, 그 이후에는 먼져 start 라는 입력값을 받을 수 있도록 처리하는 validateStartCommand 메소드를 추가해주었고, 그 이후에는 게임이 진행되는 play() 메소드를 통하여 진행되다가 "end" 가 입력될 시 종료하도록 했다. 게임 진행동안은 입력시 마다 print() 를 통해 Board 판이 출력되도록 하였다.

tmdgusya commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -20,6 +21,15 @@ public void initialize() {
     }
 
+    public void main() {
+        Scanner scanner = new Scanner(System.in);
```
이부분 try-with-resources 로 close 가 되도록 했어야 됬는데 이부분은 제가 놓친것 같습니다.
중요한건데 ㅠㅠ

ksundong replied:
셀프리뷰 좋네요 👍

ksundong replied:
수정해서 올리셔도 됩니다 ㅋㅋ

honux77 replied:
괜찮아요 ㅋㅋㅋ 어차피 스캐너 죽으면 프로그램도 끝나는데요 뭐 ㅎㅎ.

honux77 replied:
그것보다 main()이 board에 있는게 조금 마음에 걸리네요. 별도 클래스로 분리했으면 좋았을 것 같아요.

honux77 commented on src/main/java/net/honux/chess/attribute/PawnName.java:
```diff
@@ -0,0 +1,15 @@
+package net.honux.chess.attribute;
+
+public enum PawnName {
```
ㅋㅋ 이거 이름 짓기가 묘하게 어렵네요. PawnName은 뭔가 살짝 맘에 들지 않은데 딱히 더 좋은 건 저도 잘 ㅋㅋ
저라면 그냥 Representation으로 짓고, 내부 enum으로 만들었을 것 같습니다.
물론 지금 단계에서는 그냥 char로 하는게 더 좋았을 것 같아요.
현재 단계에서 enum을 사용하면 얻을 수 있는 이득이 커 보이지 않거든요.

tmdgusya replied:
char 로 수정하였습니다. 
생성자를 이용하여 색깔에 따른 representation 을 넣어주기에 생각해보니 유효성검증 부분도 필요하지 않아서,
char 를 쓰는것이 더 좋아 보이네요.

honux77 commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,30 +1,128 @@
 package net.honux.chess.core;
 
+import net.honux.chess.attribute.Color;
```
패키지 분리를 잘 하셨군요. 
역시 이런 것도 성향차이인데 저는 지금 당장은 atribute나 entity에 들어갈 대상이 하나씩이라서 이런 경우에 미리 패키지를 만들지 않습니다. 필요할 때 필요한 만큼 적용하는게 더 깔끔한 것 같아요. 순전히 개인적인 생각입니다.

tmdgusya replied:
습관이 들어서 그런지 계속하게 되더라고요.
나중에 만들어도 될 것 같긴한데.. 성격상 뭔가 정리되어서 딱딱 들어가는게 좋아보이기도 하더라고요.
근데 이렇게 작게 -> 크게로 가는 프로그래밍 방법에서는 호눅스 방법이 조금 더 효율적일것 같다는 생각이 드네요.
구현해 가면서 뭔가 계속 바뀌는 것들도 있을 것 같네요.

honux77 commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,30 +1,128 @@
 
     public void add(Pawn pawn) {
-        pawnList.add(pawn);
+        getValidateInputList(pawn).add(pawn);
```
getValidateInputList 라는 이름이 조금 모호한 느낌? 무엇을 하는건지 잘 예측되지 않네요. List로 끝나고 매개변수로 Pawn을 받으니 더욱 그렇습니다.

honux77 replied:
getListByColor() 혹은 getPawnsByColor() 정도 어땠을까요? 우리나라 분들은 List같은 컨테이너 종류를 메소드에 종종 명시하는데 영어권에서는 복수표현을 선호하는 것 같습니다.

honux77 replied:
저도 구현할 때 거의 로치랑 비슷한 코드로 구현했는데, 다른 분들은 Map<Color, List> 로 구현한 분도 있더라구요. 각각 장단점이 있는 것 같으니 참고만 하세요.

tmdgusya replied:
getListByColor 상당히 괜찮은것 같네요! 
그 K 코드 중 Map 을 이용해서 한걸 봤는데 나중에 List를 쓴걸 봤는데 나중에 Piece 에 대한 List 들이 개별로 생기면 Map이 좋을듯한 느낌도 들더라고요. 메소드도 많이 생기지 않고, 그 부분도 한번 고려해 보겠습니다. 일단은 메소드를 그냥 개별적으로 추가해주는 방식으로 진행하기는 했는데, 고려해볼만한 사항인듯 합니다

tmdgusya replied:
제 코드를 보다가 느낀건데 제가 생성자에서 representation 을 받아도 Color 에 따라 넣어주고 있었네요.
저 윗부분은 공부를 조금해보되, 지금 방식에서는 호눅스 말씀대로 Pawn 안에 char 로 넣는 것이 더 좋아보이네요

honux77 commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,30 +1,128 @@
+        return whitePawnList;
+    }
+
+    public int whitePawnsize() {
```
size -> Size

honux77 commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,30 +1,128 @@
     }
 
-    public Optional<Pawn> findPawn(int index) {
+    public Optional<Pawn> findWhitePawn(int index) {
```
우와 Optional 좋네요!

honux77 commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,30 +1,128 @@
+        try {
+            pawn = Optional.ofNullable(blackPawnList.get(index));
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("해당 위치에 Pawn 이 존재하지 않습니다.");
```
println은 나중에 좀 걸리적거리지 않을까 싶습니다. logger를 사용하는 편이 좋지 않을까요?

honux77 replied:
참 그리고 catch (e) { return Optional.empty()} 가 더 좋다고 합니다.

tmdgusya replied:
오 이게 더 깔끔한 방식인것 같네요.
Optional.empty() 를 넘겨주는 것이 좋은것 같아 보이네요
Logger 에 관한 부분은 조금 공부해보고, 적용시킬 수 있도록 해보겠습니다.

honux77 commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -1,30 +1,128 @@
+        }
+    }
+
+    private String validateStartCommand(Scanner scanner, String userInputCommand) {
```
scanner.nextLine() 때문에 스캐너를 매개변수로 받아줬는데, 메소드의 원래 역할보다 조금 더 일을 한 듯한 느낌이에요. 
스캐너 매개변수를 없애면 전체 코드가 어떻게 되는지 궁금하네요.

tmdgusya replied:
확실히 분리하니까 코드 라인도 줄어들었네요. 
메소드 이름에도 맞지 않는 느낌이였네요. 기존 코드는

honux77 commented on src/main/java/net/honux/chess/entity/pieces/Pawn.java:
```diff
@@ -1,20 +1,35 @@
         this.color = color;
+        if (color == Color.BLACK) {
+            this.representation = PawnName.BLACK_REPRESENTATION;
+        } else
```
else 제거

tmdgusya replied:
평시에는 else 가 애매하게 작동할 수 있어 좋지 않으나 이 경우에는 정확하지 않나요?
Color 에 BLACK, WHITE 말고 도 다른 속성이 추가되면 또 다시 코드를 재작성 해야 하기에 else 를 빼야하는 건가요?

honux77 commented on src/main/java/net/honux/chess/entity/pieces/Pawn.java:
```diff
@@ -1,20 +1,35 @@
+public class Pawn {
 
     private final Color color;
+    private final PawnName representation;
```
제 느낌은 역시 PawnName이 없었으면 좋았을 것 같다는 생각이.... (그냥 개인적인 소견입니다.)
코드의 복잡성만 증가시킨 것 같은데 로치 생각은 어떠세요? 
한번 `private final char representation` 으로 바꿔보면 어떨까요?

tmdgusya replied:
char 로 할경우 'c' 같은것도 들어올 수 있다고 생각이 들었고, 그 경우 유효성검사 로직을 거치는 것보다는
Enum 을 써보자 라는 생각으로 일단은 하기는 했는데, 만약 Piece 가 Pawn 이 아니라 여러개라면 코드의 복잡도를 충분히 높일 수도 있다고 생각되네요. 아직 경험상으로는 두가지 방법 중 뭐가 좋은지 판단하기 힘드네요.

honux77 commented on src/test/java/net/honux/chess/core/BoardTest.java:
```diff
@@ -22,18 +22,54 @@ void generateBoard() {
+                () -> assertThat(board.whitePawnsize()).isEqualTo(9),
+                () -> assertThat(white).isEqualTo(board.findWhitePawn(8).get()),
+                () -> assertThat(board.blackPawnsize()).isEqualTo(9),
+                () -> assertThat(black).isEqualTo(board.findBlackPawn(8).get())
```
갑자기 저도 헷갈리는데 assertThat(actual).isEqualTo(expected) 인 것 같은데 순서를 거꾸로 한 것 같아요. 확인요망

tmdgusya replied:
아 거꾸로 한것 같습니다.
수정하도록 하겠습니다 :)

honux77 commented on src/test/java/net/honux/chess/entity/pieces/PawnTest.java:
```diff
@@ -1,36 +1,48 @@
-        Color expectedColor = Color.BLACK;
-        Color result = pawn.getColor();
-        assertThat(result).isEqualTo(expectedColor);
+        Assertions.assertAll(
```
요것도 코드 중복인데 제거 가능하지 않을까요?

honux77 reviewed (changes requested):
전반적으로 잘 짰는데 한 번 더 고민해 보고 조금만 개선해 보세요! 수고하셨습니다! 역시 로치!

tmdgusya committed:
reivew 3차 미션 코드리뷰 반영

Board.java
- PawnSize() 부분 CamelCase 로 수정
- findPawn() 부분 에러메세지보다는 Optional.empty() 리턴 하도록 수정하였음.
- main() 주요 동작 부분 ChessGame.java 로 분리하였음 -> 관심사의 분리가 필요하다고 판단되었음

BoardTest.java

- expect actual 반대로 쓴 부분들 수정

ChessGame.java

- 유저 입력 및 인터페이스 부분 최대한 else 문을 안쓰기 위해 else 제거하였음.
- validateStartCommand 의 메소드가 지니는 의미는 유효성검사의 책임만 지면 된다고 판단되어 해당 메소드는 유효성 검사만 진행하도록 했음

PawnName.java : 필요없다고 판단되어 삭제하였음.

Pawn.java

- 생성자에서 else 를 덜어내기 위해서 지역변수인 defaultRepresentation 를 추가해서 default 는 하얀색이므로, BLACK 이 아니면 하얀색을 넣도록 했음. else 를 덜어내자니 final 이 발을 붙잡아서 지역변수를 이용하였음.

tmdgusya commented:
@honux77 

review 커밋 반영해서 다시 푸시했습니다.

else 를 덜어내는 것은 익숙하지 않다보니 뭔가 시간이 길어졌던것 같네요. 로직적으로도 뭔가 줄이 길어지려 하기에 그걸 최대한 줄여보려고 노력했습니다. 네이밍은 참 어렵다는 생각이 듭니다. 어서 영어 공부를.. 하거나 남의 코드를 많이 봐야겠다는 생각이 드네요.
